rust/diesel/src/order.rs

```diff
@@ -109,16 +115,29 @@ impl Order {
                     .execute(connection)
             });
 
-        update_result.map(|_| order).ok()
+        match update_result {
+            Err(diesel::result::Error::NotFound) | Ok(0) => Ok(None),
+            Err(err) => Err(err.to_string()),
+            Ok(_) => Ok(Some(order)),
+        }
     }
 
-    fn create_new_order(order: &NewUserOrder, connection: &PgConnection) -> Option<(NewOrder, Vec<(Product, i16)>)> {
+    fn create_new_order(
+        order: &NewUserOrder,
+        connection: &PgConnection,
+    ) -> Result<(NewOrder, Vec<(Product, i16)>), String> {
         let product_units: Vec<(Product, i16)> = order
             .products
             .iter()
             .flat_map(|product_ref| {
-                Product::read(product_ref.product_id, connection)
-                    .map(|product| (product, product_ref.units))
+                let ret: Vec<(Product, i16)> = Product::read(product_ref.product_id, connection)
```

Comment:
Errors that occur when reading the product are being ignored, I don't think that is what we want.
I think you could use something along these lines here:
```
let product_units: Vec<(Product, i16)> = order
            .products
            .iter()
            .map(|product_ref| {
                match Product::read(product_ref.product_id, connection)? {
                    None => Err(Box::new(DbError::NotFound)),
                    Some(product) => Ok((product, product_ref.units)),
                }
            })
            .collect()?;
```